### PR TITLE
[Offline editing] Unbreak TestQgsOfflineEditing with GDAL >= 3.4.2

### LINKED ISSUE #48012: ci testsuite: TestQgsOfflineEditing fails with gdal >= 3.4.2
### What is the bug or the crash?
test TestQgsOfflineEditing fails with gdal 3.4.2 related to https://github.com/qgis/QGIS/commit/fe410ec92051b5be1d73b5753f05a7cfdc004112
```
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
FAIL!  : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() Compared values are not the same
   Actual   (gpkgLayer->name())                        : "points_gpkg"
   Expected (QStringLiteral( "points_gpkg (offline)" )): "points_gpkg (offline)"
   Loc: [tests/src/core/testqgsofflineediting.cpp(299)]
```

### Steps to reproduce the issue

compile & test master with gdal 3.4.2 (i.e. on fedora 36 as on PR #47887)

details see https://cdash.orfeo-toolbox.org/testDetails.php?test=35806854&build=95038

### Versions

master with gdal=3.4.2

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

https://github.com/qgis/QGIS/blob/20dfa3c038a3325c9b17e92a05c2d887e73a1427/tests/src/core/testqgsofflineediting.cpp#L296-L299

```
********* Start testing of TestQgsOfflineEditing *********
Config: Using QtTest library 5.15.2, Qt 5.15.2 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.2.0), gentoo unknown
PASS   : TestQgsOfflineEditing::initTestCase()
PASS   : TestQgsOfflineEditing::createSpatialiteAndSynchronizeBack(no suffix)
PASS   : TestQgsOfflineEditing::createSpatialiteAndSynchronizeBack(null suffix)
PASS   : TestQgsOfflineEditing::createSpatialiteAndSynchronizeBack(empty suffix)
PASS   : TestQgsOfflineEditing::createSpatialiteAndSynchronizeBack(part of name suffix)
PASS   : TestQgsOfflineEditing::createSpatialiteAndSynchronizeBack(another suffix)
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(no suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(no suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(no suffix) FID: 17 Class: "Superjet"
PASS   : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(no suffix)
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix) FID: 17 Class: "Superjet"
PASS   : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(null suffix)
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix) FID: 17 Class: "Superjet"
PASS   : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(empty suffix)
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix) FID: 17 Class: "Superjet"
PASS   : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(part of name suffix)
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix) FID: 17 Class: "Superjet"
PASS   : TestQgsOfflineEditing::createGeopackageAndSynchronizeBack(another suffix)
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffNames' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
QDEBUG : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() "Offline Editing Plugin" "Field 'StaffAges' from layer points has been converted from a list to a JSON-formatted string value."
FAIL!  : TestQgsOfflineEditing::removeConstraintsOnDefaultValues() Compared values are not the same
   Actual   (gpkgLayer->name())                        : "points_gpkg"
   Expected (QStringLiteral( "points_gpkg (offline)" )): "points_gpkg (offline)"
   Loc: [tests/src/core/testqgsofflineediting.cpp(299)]
PASS   : TestQgsOfflineEditing::cleanupTestCase()
Totals: 12 passed, 1 failed, 0 skipped, 0 blacklisted, 870ms
********* Finished testing of TestQgsOfflineEditing *********
```

### PR DESCRIPTION
**Pull request by rouault**
Fixes #48012

Basically, for offline editing, we ask QgsOgrProvider to adopt the
behaviour it had before https://github.com/qgis/QGIS/pull/47098, that is
use SQLite3 journal_mode = WAL even in read-only mode.

Sorry, this "fix" is horrible, but I couldn't get with something better after
many hours of investigation. The root cause seems to be that QgsOfflineEditing
creates layers using direct OGR API or sqlite3 API, and also uses
QgsOgrProvider which has a GDAL dataset reuse mechanism. And all that does
not place nicely at all. Maybe the root root cause is that the OGR GPKG driver
establishes the list of layers at its opening, and if creates a new one behind its
back, which is prone to happen with the dataset caching metchanism, the
GDALDataset::GetLayerByName() method doesn't see the newly created layer.

With that in mind, I tried to add calls to
QgsOgrProviderUtils::invalidateCachedDatasets() but to no avail.
Also QgsOgrProviderUtils::invalidateCachedLastModifiedDate() which is
used by QgsOgrProvider::createEmptyLayer(), but to no avail.

So basically I don't understand why offline editing worked before with a
GPKG backend. I'd say it is pure luck.

Note: I also saw that qgsofflineediting.cpp at line 315 calls
offlineLayer->dataProvider()->invalidateConnections() but that method is
only implemented in the Spatialite provider. I tried to implement it in
the OGR one, but to no avail too.

This should probably be backported, despite being horrible...

### CODE REVIEW

**Comment by rouault** [comment]
I've also added a second commit "[OGR provider] (Presumably) fix opening of GeoPackage from network share on Windows with GDAL >= 3.4.2 (fixes #48003)"  (I couldn't test its effect not having the needed setup). It is in the same spirit as the first one: for non-local files, fallback to the old method not using NOLOCK=YES and SQLite3 URIs since I suspect that network share filenames to not be compatible of them

**Comment by agiudiceandrea** [comment]
> This should probably be backported

At least for the second commit "[OGR provider] (Presumably) fix opening of GeoPackage from network share on Windows with GDAL >= 3.4.2 (fixes https://github.com/qgis/QGIS/issues/48003)", I think this PR needs to be backported to 3.24 and to 3.22 before the release 3.22.6 which will be otherwise affected by the https://github.com/qgis/QGIS/issues/48024 regression.

**Review by nyalldawson — APPROVED** [review]

**Comment by agiudiceandrea** [comment]
> I've also added a second commit "[OGR provider] (Presumably) fix opening of GeoPackage from network share on Windows with GDAL >= 3.4.2 (fixes #48003)" (I couldn't test its effect not having the needed setup). It is in the same spirit as the first one: for non-local files, fallback to the old method not using NOLOCK=YES and SQLite3 URIs since I suspect that network share filenames to not be compatible of them

@rouault I've tested QGIS 3.25.0-Master (dd6e83b2fa) from OSGeo4W (with GDAL 3.4.2) and it seems to me the issue is not yet fixed.

Trying to add a gpkg layer from a network location dragging and dropping it in a map, generates the following errors:
```
CRITICAL Invalid Data Source : //192.168.1.1/test/test.gpkg is not a valid or recognized data source.
WARNING Unable to load //192.168.1.1/test/test.gpkg
```
and the layer is not added.

Trying to add a gpkg layer from a network location using the Browser panel the following error is displayed:
```
The file does not contain any layer or there was an error opening the file. Check file and directory permissions on \\192.168.1.1\test\test.gpkg
```
and the layer is not added.

The layer can only be added using the Data Source Manager.
